### PR TITLE
fix(json): define JSONL as one record per physical line on every transport

### LINKED ISSUE #486: Define and enforce JSONL record boundaries consistently across transports
## Dogfooding evidence

The file and async scanners treat JSONL as a stream of top-level JSON values, while synchronous bytes split on lines. The same payload therefore has incompatible strict/tolerant behavior:

```python
payload = b'{"x":1}{"x":2}'
```

- file, skip: 2 rows, `error_count == 0`
- file, strict: 2 rows, `error_count == 0`
- bytes, skip: fails because there are no valid lines
- bytes, strict: raises `ValueError`
- async follows the file scanner

With `b'{"x":1}{"x":2}\n{"x":3}\n'`, file/async report 3 clean rows while bytes skip reports 1 row and 1 error.

Space-separated values on one physical line show the same mismatch.

## Why this needs a product decision

README/API language says JSONL is “one object per line,” which favors a line-oriented parser. However, `dataprof-json` intentionally tests a pretty-printed multi-line root object as one JSONL row, and automatic JSON detection routes any non-array root through the JSONL scanner. Making JSONL strictly line-based would break that supported single-object behavior unless standard JSON object input gets its own explicit parser mode.

This is the remaining boundary-semantics gap behind #382, related to the JSON path consistency work in #463 and source-order work in #465.

## Acceptance criteria

- Decide and document whether `format="jsonl"` means strict physical lines or a whitespace-delimited stream of JSON values.
- Give standard JSON arrays, standard JSON objects, and JSONL distinct unambiguous parser modes/hints.
- File, bytes, and async transports apply identical record-boundary and strict/tolerant semantics.
- Adjacent top-level values without an allowed delimiter never produce a clean profile.
- Preserve a supported path for compact and pretty-printed single JSON objects if that remains part of the product contract.
- Add parity tests for adjacent values, same-line whitespace separation, pretty-printed objects, blank lines, and malformed first/middle/last records.
- Update README/API docs to match the chosen grammar.

### PR DESCRIPTION
**Pull request by AndreaBozzo**
Closes #486.

## The decision

#486 asked for a product decision before code, and this is it:

- **`format="jsonl"` — one record per physical line.** A record may not span
  lines; a line may not hold more than one value. Blank lines are separators,
  neither records nor errors.
- **`format="json"` — one standard JSON document.** An array of objects, *or* a
  single object as one record. Whitespace is insignificant, so either may be
  pretty-printed across lines.

Whitespace is insignificant in one grammar and load-bearing in the other, so an
input cannot be valid under both and mean the same thing. That is what makes the
two unambiguous.

## What was wrong

The file and async scanners read JSONL as a whitespace-delimited stream of
top-level JSON values; the synchronous bytes reader split on lines. Same payload,
incompatible answers — verified on master before touching anything:

| input, as `.jsonl` | file & async | bytes |
| --- | --- | --- |
| `{"x":1}{"x":2}` | 2 rows, 0 errors | rejected |
| `{"x":1}{"x":2}\n{"x":3}\n` | 3 rows, 0 errors | 1 row, 1 error |
| `{"x":1} {"x":2}\n` | 2 rows, 0 errors | rejected |
| pretty-printed object | 1 row, 0 errors | rejected |

The first row is the one that matters: a concatenation whose delimiter was lost
profiled as two clean rows. Silently reading damaged data as clean is the
failure a profiler exists to catch.

## What changed

`JsonFormat` is now `Json` (standard document, array or single object) and
`Jsonl` (strict lines). The JSONL branch of both scanners reads a line at a time
and parses it with `from_str`, which requires the value to fill the input — so a
line holding two values is one malformed record, not two clean ones. The JSON
branch gained the single-object shape, which is what lets pretty-printed
documents work without JSONL having to accept multi-line records.

A leading `{` opens both grammars, so sniffing cannot separate them and the
caller says which it has: the facade passes the resolved `FileFormat`, and
`analyze_json_file` falls back to the file extension. A leading `[` is still
resolved from content.

**Two latent bugs surfaced on the way, both fixed here:**

1. The sync array reader called `reader.consume(1)` after
   `consume_leading_whitespace`, which only *peeks*. With whitespace before `]`
   it ate a newline and left the bracket, which then looked like content after
   the closing bracket — so **every pretty-printed JSON array reported a phantom
   `error_count = 1`**. Confirmed pre-existing on master, not introduced here.
2. Per-line parsing makes the decoder's own line number always 1. Malformed
   JSONL records are now located by their physical line, matching what the
   Python bytes reader already did.

## Verification

`python/tests/test_json_record_boundaries.py` (24 tests) checks every case on
**all four transports at once** — file, bytes, async file, async bytes — because
the bug was never visible from any single one. Cases cover adjacent values,
same-line whitespace separation, pretty-printed objects and arrays, blank lines,
malformed first/middle/last records, truncated records, fieldless records, and a
5,000-row stream that crosses chunk boundaries.

**13 of the 24 fail against unfixed code.** The 11 that pass are the
already-correct cases the change must preserve.

A full sweep after the change shows 0 mismatches across 16 payloads × 4
transports × both grammars.

## Behaviour changes to be aware of

- A `.jsonl` file with adjacent or space-separated values on one line is now an
  error (skip mode: counted; strict: raised). **This is the point of the change**,
  but it will fail inputs that previously profiled clean.
- A pretty-printed object read as `.jsonl` is now rejected; read as `.json` it is
  one record. The `dataprof-json` test that asserted the old "pretty object is
  one JSONL row" behaviour is replaced by tests for both halves of the new rule.
- `JsonFormat::JsonArray` → `JsonFormat::Json`, and
  `JsonParserConfig::json_array()` → `json_document()`, now accepting a bare
  object as well as an array.
- Pretty-printed JSON arrays report `error_count = 0` instead of `1`.

## Commands run

```
cargo fmt --all
cargo clippy --all --all-targets --all-features -- -D warnings
cargo test --workspace --all-features          # 39 suites, 829 tests
uv run maturin develop --features "python,python-async,async-streaming,database,sqlite"
uv run ruff format / ruff check python/ .github/scripts/ .claude/skills/dataprof/scripts/
uv run ty check python/
uv run pytest python/tests/                    # 936 passed, 4 skipped, 1 xfailed
```

### CODE REVIEW

File: crates/dataprof-engines/src/streaming/async_reader.rs [hunk 1]
```diff
@@ -63,6 +63,16 @@ fn peek_non_whitespace<R: std::io::BufRead>(
...
 }
 
+/// Build a strict-mode error for a standard JSON document that is not exactly
+/// one value. Mirrors `dataprof_json::json_document_error`.
+fn json_document_error(err: &serde_json::Error) -> DataProfilerError {
+    DataProfilerError::JsonParsingError {
+        message: format!(
+            "malformed JSON document: {err}. A JSON source must hold exactly one              array or object; for one record per line use format=\"jsonl\""
+        ),
+    }
```
[the remark this answers is not in the document]
Comment (AndreaBozzo) [reply crates/dataprof-engines/src/streaming/async_reader.rs:74]: Confirmed and fixed in 597d562 — good catch, this was a real corruption.

Root cause is nastier than it looks: `cargo fmt` **joined the `\`-continued line and left its indentation inside the string**. So the literal in the source no longer said what it appeared to say, and only the async copies were affected because their deeper indentation pushed them past the width where rustfmt reflows. Rendered output before the fix:

```
sync : ... must hold exactly one array or object; ...
async: ... must hold exactly one              array or object; ...
```

The second message you flagged (line 830) had the same fault, worse — a ~40-space run.

Both are now single source lines, which rustfmt does not split, with a comment saying why. Verified by rendering the real errors rather than reading the source: sync and async are now byte-identical.

File: crates/dataprof/src/profiler.rs [hunk 2]
```diff
@@ -569,6 +569,19 @@ impl Profiler {
     }
 
     /// A JSON parser config carrying the configured row cap and error policy.
+    /// A JSON parser config for `format`, carrying the row cap and error policy.
```
[the remark this answers is not in the document]
Comment (AndreaBozzo) [reply crates/dataprof/src/profiler.rs]: Fixed in 597d562. You are right about the cause: inserting `json_config_for` above `json_config_for_stop` left the latter's summary line stranded on top of the new one. `json_config_for_stop` has its own doc comment back.

File: crates/dataprof-json/src/lib.rs [hunk 3]
```diff
@@ -123,83 +152,70 @@ where
...
-                if found_array || consume == 0 {
-                    break;
-                }
             }
+        }
+        FileFormat::Json => {
+            // Which of the two record-carrying shapes this document is. The
+            // first non-whitespace byte decides, and is left unconsumed so the
+            // object branch still sees the value it opens.
+            let found_array = consume_leading_whitespace(&mut reader)? == Some(b'[');
```
[the remark this answers is not in the document]
Comment (AndreaBozzo) [reply crates/dataprof-json/src/lib.rs:206]: Half right — the truncation reason is genuinely lost, but not for the reason given, and not only for single-object documents.

I checked all three shapes with `max_rows=0` against a non-empty source:

```
jsonl 2 records   rows=0 truncation=None
json array        rows=0 truncation=None
json object       rows=0 truncation=None
```

The JSONL and array paths *do* set `truncated` correctly. It is dropped later, by the `rows_read == 0` branch in `analyze_json_file_with_options`, which returns an empty report before any truncation reason is applied. So `scan_single_json_object` returning `(0, 0)` is not the cause — passing `truncated` out of it would change nothing.

Since this predates the PR and affects every shape (and possibly CSV/Parquet too), I have not folded a fix in here — filed as #541 with the acceptance criteria, including checking the other formats before fixing so it lands as one contract.

What I did fix in 597d562 is the comment in `scan_single_json_object`, which claimed the read *is* reported as truncated. It now says plainly that it is not, and points at the shared cause.
